File: Editorial/111_Minimum_Depth_of_Binary_Tree/Python3/2_Breadth_First_Search.py
class Solution:
    def minDepth(self, root: TreeNode) -> int:
        if not root:
            return 0
        q = collections.deque([root])
        depth = 1
        while q:
            qSize = len(q)
            for _ in range(qSize):
                node = q.popleft()
                # Since we added nodes without checking null, we need to skip them here.
                if not node:
                    continue
                # The first leaf would be at minimum depth, hence return it.
                if not node.left and not node.right:
                    return depth
                q.append(node.left)
                q.append(node.right)
            depth += 1
        return -1

File: Editorial/111_Minimum_Depth_of_Binary_Tree/TypeScript/2_Breadth_First_Search.ts
function minDepth(root: TreeNode | null): number {
    if (!root) {
        return 0;
    }
    let q: (TreeNode | null)[] = [root];
    let depth: number = 1;
    while (q.length) {
        let qSize: number = q.length;
        for (let i = 0; i < qSize; i++) {
            let node: TreeNode | null = q.shift()!;
            // Since we added nodes without checking null, we need to skip them here.
            if (!node) {
                continue;
            }
            // The first leaf would be at minimum depth, hence return it.
            if (node.left === null && node.right === null) {
                return depth;
            }
            q.push(node.left);
            q.push(node.right);
        }
        depth++;
    }
    return -1;
}
